### PR TITLE
Implement backup of empty dirs efficiently via new `FileTree` approach

### LINKED ISSUE #4: Handle directories
Should really copy empty directories... nice to have it.
- Essentially copy leaves (files & empty dirs)
  - Checking for empty dir may be expensive however. Don't think `attributes` has this info.
- Create parent dirs as it does now

There's also the issue of not cleaning up directories. **Solved on #5**
- E.g. backed up `A/B/C.pdf`
  - Then delete `A/B/C.pdf`... 
  - On next run directories `A/B` remain on destination! (since only files are scanned)
- So when going over destination we need to attempt deletion of any path.
  - Most will result in 'not found' exception since we do recursive delete. But that's fine. Just ignore. 
  - Leaves would not be enough. 
    - In above example `A/B` would be deleted but `A` would remain. 
    - I guess on the next run `A` would be deleted... so eventually stuff gets cleaned up.
- Don't think unit test to cover this is possible, since it involves scan, run, delete all together
  - So add a set of integration tests
- How does AWS S3 handle this scenario? List object doesn't return folders right? Might return empty folders?

### PR DESCRIPTION
**Pull request by will-molloy**
_Follow-up to PR #5 and solves issue #4_

### Implement backup of empty dirs efficiently via new `FileTree` approach
- Basically no need to test for empty dir
- Instead attempt `put` with every path, but skips if it would be covered by a descendant
  - Similarly for `delete`, except skips if it would be covered by an ancestor
- This was the original idea btw: https://github.com/will-molloy/file-backup/commit/d6c09d91dbb01c0a3fd352d894f09be5c2ba67e1#diff-1e9073ebc5be97e216ce1d6552c1c2edf8f34680179585662ee76976715d077fR40-R41
  - But gave up since it was O(n^2).
- Implemented efficiently here via `Path` based `Trie`, abstracted behind `FileTree` interface
  - Runtime is ~O(n*p) - n = number of nodes, p = average path length

### Quite a bit of refactoring, improvement of test coverage and bug fixes (found by the new tests).
- Main bug fix was overwriting file with non-empty dir in `LocalBackup`
  - Fun with [`Files.createDirectories`](https://stackoverflow.com/a/76278968/6122976)

### Removed `Statistics`. 
- It was getting too complicated; would need `put`/`delete` to report `Statistics` to do it properly
- Also not used when running backups on schedule...

### CODE REVIEW

File: file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java [hunk 1]
```diff
@@ -94,6 +95,17 @@ private OverallStatistics run() {
...
+            // TODO WIP - minimal delete...
+            // TODO just make scan key a Path?
+            int parentI = key.lastIndexOf('/');
+            if (parentI > 0){
+              String parent = key.substring(0, key.lastIndexOf('/'));
+              if (destNodes.containsKey(parent) && !sourceNodes.containsKey(parent)){
+                // skip... deletion of parent will cover this node
+                return;
+              }
+            }
```
Comment (will-molloy) [inline file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java]: If we can do something similar for create... then it would have solved the problem of needing to check for an empty directory.

___

However can't simply add (children) to the path. That's adding unknown info.

Could do a prefix check? But that's O(1) -> O(n) (and O(n) -> O(n^2) overall)... unless we use a different structure.

File: file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java [hunk 2]
```diff
@@ -94,6 +95,17 @@ private OverallStatistics run() {
       // otherwise empty dirs are left on destination
       destNodes.forEach(
           (key, destFile) -> {
+            // TODO WIP - minimal delete...
+            // TODO just make scan key a Path?
+            int parentI = key.lastIndexOf('/');
+            if (parentI > 0){
+              String parent = key.substring(0, key.lastIndexOf('/'));
+              if (destNodes.containsKey(parent) && !sourceNodes.containsKey(parent)){
+                // skip... deletion of parent will cover this node
```
Comment (will-molloy) [inline file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java]: This messes up the delete counter.

Need to return number of nodes deleted?

File: file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java (hunk 1 above)
Comment (will-molloy) [reply file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java]: https://stackoverflow.com/questions/46625774/finding-entries-in-a-set-containing-a-prefix

File: file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java (hunk 2 above)
Comment (will-molloy) [reply file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java]: Or return `Set<Path>` that was deleted?

File: file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java [hunk 3]
```diff
@@ -46,51 +47,55 @@ public S3Bucket destination() {
   }
 
   @Override
-  public boolean put(String key) {
+  public boolean put(Path key) {
```
Comment (will-molloy) [inline file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java]: Need to test S3 put with empty dir.

File: file-backup-core/src/main/java/com/willmolloy/backup/TrieBasedFileTree.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+      }
+
+      Stream<Node> stream() {
+        return Stream.concat(Stream.of(this), children.values().stream().flatMap(Node::stream))
+            .filter(Node::containsData);
+      }
+
+      boolean containsData() {
+        return path != null;
+      }
```
Comment (will-molloy) [inline file-backup-core/src/main/java/com/willmolloy/backup/TrieBasedFileTree.java]: Just assume every node contains data?

File: file-backup-local/src/main/java/com/willmolloy/backup/local/LocalFile.java [hunk 5]
```diff
@@ -38,18 +42,23 @@ public long size() {
...
+  @Override
+  public boolean isDirectory() {
+    return isDirectory;
+  }
+
   @Override
   public boolean same(File other) {
     if (other instanceof LocalFile localFile) {
-      return size() == other.size() && lastModified().equals(localFile.lastModified());
+      return File.super.same(other) && lastModified.equals(localFile.lastModified);
```
Comment (will-molloy) [inline file-backup-local/src/main/java/com/willmolloy/backup/local/LocalFile.java]: This doesnt seem to work? It's running put on everything??

File: file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java [hunk 6]
```diff
@@ -46,51 +47,55 @@ public S3Bucket destination() {
...
+  public boolean delete(Path key) {
     String destinationUri = destination.objectUri(key);
     try {
       DeleteObjectRequest request =
           DeleteObjectRequest.builder()
               .bucket(destination.bucketName())
-              .key(destination.prefix() + key)
+              .key(ensureUnixSeparator(destination.prefix().resolve(key)))
               .build();
       s3Client.deleteObject(request);
```
Comment (will-molloy) [inline file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java]: Should add a waiter, described here https://www.codejava.net/aws/create-folder-examples

File: file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java [hunk 7]
```diff
@@ -46,51 +47,55 @@ public S3Bucket destination() {
...
     } catch (RuntimeException | IOException e) {
-      log.error("Error putting: [%s] -> [%s]".formatted(sourcePath, destinationUri), e);
+      log.error("Error putting: [{}] -> [{}]", sourcePath, destinationUri, e);
       return false;
     }
   }
 
   @Override
-  public boolean delete(String key) {
+  public boolean delete(Path key) {
```
Comment (will-molloy) [inline file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java]: This also needs to handle dirs... need the "/"?

File: file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java (hunk 6 above)
Comment (will-molloy) [reply file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java]: Do we need it??? Do it later anyway...

File: file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java (hunk 3 above)
Comment (will-molloy) [reply file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java]: Working now... but delete is not :)

File: file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java [hunk 8]
```diff
@@ -46,51 +52,69 @@ public S3Bucket destination() {
...
     }
   }
 
   @Override
-  public boolean delete(String key) {
-    String destinationUri = destination.objectUri(key);
+  public boolean delete(S3File destFile) {
+    Path key = destFile.relativePath();
     try {
+      // TODO DeleteObjectsRequest for folders... currently does not work.
```
Comment (will-molloy) [inline file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java]: Needs resolving

Comment (will-molloy) [reply file-backup-s3/src/main/java/com/willmolloy/backup/s3/S3Backup.java]: It deletes objects and empty dirs but not non-empty dirs

File: file-backup-local/src/main/java/com/willmolloy/backup/local/LocalFile.java (hunk 5 above)
Comment (will-molloy) [reply file-backup-local/src/main/java/com/willmolloy/backup/local/LocalFile.java]: 9b9cc0837f0d675a0cad2e74d9f8ec6e9fe0990d

File: file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java [hunk 9]
```diff
@@ -32,228 +27,107 @@
...
+        log.error("Interrupted", e);
+        Thread.currentThread().interrupt();
       }
-    } else {
-      failedDeleteCount.incrementAndGet();
     }
+    return allSuccess.get();
   }
 
-  private OverallStatistics getAndLogStats(long runStartNanos) {
```
Comment (will-molloy) [inline file-backup-core/src/main/java/com/willmolloy/backup/BackupRunner.java:155]: Could add a simple statistics? Put/Delete/Same count?

Issue was mainly with the bytes added/removed and delete count.
